File: operator/pkg/controllers/psstrategy/worker.go
// Copyright 2022 The EasyDL Authors. All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package psstrategy

import (
	"context"
	elasticv1alpha1 "github.com/intelligent-machine-learning/easydl/operator/api/v1alpha1"
	controllers "github.com/intelligent-machine-learning/easydl/operator/pkg/controllers"
	logger "github.com/sirupsen/logrus"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	"sort"
	"strconv"
)

const (
	workerServicePort int = 2222
)

// WorkerManager generates a master pod object.
type WorkerManager struct {
	PSTaskManager
}

func init() {
	logger.Infof("init worker manager")
	controllers.ReplicaManagers[ReplicaTypeWorker] = newWorkerManager()
}

func newWorkerManager() *WorkerManager {
	return &WorkerManager{
		PSTaskManager: PSTaskManager{
			taskType: ReplicaTypeWorker,
		},
	}
}

// ReconcilePods creates a Pod on a K8s cluster
func (m *WorkerManager) ReconcilePods(
	r *controllers.ElasticJobReconciler,
	job *elasticv1alpha1.ElasticJob,
	resourceSpec *elasticv1alpha1.ReplicaResourceSpec,
) error {
	workerStatus := m.getTaskStatus(job)
	currentNum := m.getTotalTaskCount(workerStatus)
	aliveNum := int(workerStatus.Active + workerStatus.Pending)
	if resourceSpec.Replicas > aliveNum {
		m.scaleUpWorkers(r, job, currentNum, resourceSpec.Replicas-aliveNum)
	} else {
		m.scaleDownWorkers(r, job, aliveNum-resourceSpec.Replicas)
	}
	return nil
}

func (m *WorkerManager) scaleUpWorkers(
	r *controllers.ElasticJobReconciler,
	job *elasticv1alpha1.ElasticJob,
	currentNum int,
	upNum int,
) error {
	cluster := m.getPSCluster(r.Client, job)
	for i := currentNum; i < currentNum+upNum; i++ {
		cluster.Worker[i] = m.newTaskServiceAddr(job.Name, i, workerServicePort)
		worker := m.newTask(job, i)
		m.insertTfConfigToEnv(&worker.Spec.Containers[0], cluster, i)
		err := r.Create(context.Background(), worker)
		if err != nil {
			r.Recorder.Eventf(
				job,
				corev1.EventTypeWarning,
				string(corev1.PodFailed),
				"Worker pod %s created failed: %v",
				worker.Name,
				err,
			)
			return err
		}
		service := m.newTaskService(job, i, workerServicePort)
		err = r.Create(context.Background(), service)
		if err != nil {
			r.Recorder.Eventf(
				job,
				corev1.EventTypeWarning,
				string(corev1.PodFailed),
				"worker service %s created failed: %v",
				service.Name,
				err,
			)
			return err
		}
	}
	return nil
}

func (m *WorkerManager) scaleDownWorkers(
	r *controllers.ElasticJobReconciler,
	job *elasticv1alpha1.ElasticJob,
	downNum int,
) error {
	workers, err := m.GetReplicaTypePods(r.Client, job, m.taskType)
	if errors.IsNotFound(err) {
		logger.Warningf("No any worker found: %v", err)
		return nil
	}
	aliveWorkers := make(map[int]*corev1.Pod)
	workerIndices := []int{}
	for _, worker := range workers {
		if worker.Status.Phase == corev1.PodRunning || worker.Status.Phase == corev1.PodPending {
			workerIndex, _ := strconv.Atoi(
				worker.Labels[controllers.LabelReplicaIndexKey],
			)
			workerIndices = append(workerIndices, workerIndex)
			aliveWorkers[workerIndex] = &worker
		}
	}
	sort.Sort(sort.Reverse(sort.IntSlice(workerIndices)))
	for i := 0; i < downNum; i++ {
		m.DeletePod(r.Client, job, aliveWorkers[i])
	}

	return nil
}

File: dlrover/python/elastic_agent/common/worker.py
# Copyright 2024 The DLRover Authors. All rights reserved.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from torch.distributed.elastic.agent.server.api import RunResult, WorkerSpec
from dlrover.python.diagnosis.common.diagnose_action import (
    DiagnoseActionQueue,
    DiagnoseAction,
)
from typing import Optional, List


class WorkerContext:
    def __init__(self):
        self._worker_spec: Optional[WorkerSpec] = None
        self.remaining_failovers = 0
        self.restart_count = 0
        self._run_result: Optional[RunResult] = None
        self._diagnose_action_queue = DiagnoseActionQueue()

    @property
    def worker_spec(self):
        return self._worker_spec

    @property
    def run_result(self):
        return self._run_result

    def to_string(self) -> str:
        return (
            "WorkerContext:\n"
            f"worker_spec: {self._worker_spec}\n"
            f"remaining_failover: {self.remaining_failovers}\n"
            f"restart_count: {self.restart_count}\n"
            f"run_result: {self._run_result}"
        )

    def update_context(
            self,
            worker_spec: WorkerSpec = None,
            remaining_failovers: int = 0,
            restart_count: int = 0,
            run_result: RunResult = None,
    ):
        self._worker_spec: WorkerSpec = worker_spec
        self.remaining_failovers = remaining_failovers
        self.restart_count = restart_count
        self._run_result = run_result

    def enqueue_diagnose_action(self, action: DiagnoseAction):
        self._diagnose_action_queue.add_action(action)

    def next_actions(self) -> List[DiagnoseAction]:
        return self._diagnose_action_queue.next_actions()
